### PR TITLE
[Merged by Bors] - feat(CategoryTheory): calculus of fractions

### PR DESCRIPTION
**Pull request by joelriou**
This PR introduces notions of left and right fractions for a morphism property in a category. The typeclasses `MorphismProperty.HasLeftCalculusOfFractions` and `MorphismProperty.HasRightCalculusOfFractions` are also introduced.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    W.LeftFraction X Y := mk (𝟙 X) s hs
+
+/-- If `φ : W.LeftFraction X Y` and `L` is a functor which inverts `W`, this is the
+induced morphism `L.obj X ⟶ L.obj Y`  -/
+noncomputable def map (φ : W.LeftFraction X Y) (L : C ⥤ D) (hL : W.IsInvertedBy L) :
+    L.obj X ⟶ L.obj Y :=
+  have := hL _ φ.hs
+  L.map φ.f ≫ inv (L.map φ.s)
+
+@[reassoc (attr := simp, nolint unusedHavesSuffices)]
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:79]: This nolint points at a bug. Is there already an issue for that?

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+namespace Functor
+
+lemma congr_map_conjugate {C D : Type _} [Category C] [Category D] {F₁ F₂ : C ⥤ D}
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:32]: I guess this lemma could sensibly move to some basic file on functors?

Comment (joelriou) [reply Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean]: Thanks! Indeed, this lemma shall be used only in later PRs. I have removed it for now.

**Comment by jcommelin** [comment]
Thanks :tada: 

If CI passes, please remove the label `awaiting-CI` and merge this yourself, by adding a comment `bors r+`.

bors d+

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean]: Using `letI` instead of `have` fixed the problem...

**Comment by joelriou** [comment]
bors merge

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+/-- A multiplicative morphism property `W` has left calculus of fractions if
+any right fraction can be turned into a left fraction and that two morphisms
+that can be equalized by precomposition with a morphism in `W` can also
+be equalized by postcomposition with a morphism in `W`. -/
+class HasLeftCalculusOfFractions extends W.IsMultiplicative : Prop where
+  exists_leftFraction ⦃X Y : C⦄ (φ : W.RightFraction X Y) :
+    ∃ (ψ : W.LeftFraction X Y), φ.f ≫ ψ.s = φ.s ≫ ψ.f
+  ext : ∀ ⦃X' X Y : C⦄ (f₁ f₂ : X ⟶ Y) (s : X' ⟶ X) (_ : W s)
+    (_ : s ≫ f₁ = s ≫ f₂), ∃ (Y' : C) (t : Y ⟶ Y') (_ : W t), f₁ ≫ t = f₂ ≫ t
```
Comment (alreadydone) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:181]: Why not
```suggestion
  ext : ∀ ⦃X' X Y : C⦄ (f₁ f₂ : X ⟶ Y) (s : X' ⟶ X), 
    W s → s ≫ f₁ = s ≫ f₂ → ∃ (Y' : C) (t : Y ⟶ Y'), W t ∧ f₁ ≫ t = f₂ ≫ t
```

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+namespace MorphismProperty
+
+variable {C D : Type _} [Category C] [Category D]
```
Comment (alreadydone) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:32]: I think `Type*` is preferred now

**Pull request closed without merging**